src/lib/run-code.ts

```
@@ -0,0 +1,78 @@
+import * as esbuild from "esbuild-wasm";
+
+const supportedJS = ["js", "javascript"];
+const supportedTS = ["ts", "typescript"];
+const supportedLanguages = [...supportedJS, ...supportedTS];
+
+function isJavaScript(language: string) {
+  return supportedJS.includes(language);
+}
+
+function isTypeScript(language: string) {
+  return supportedTS.includes(language);
+}
+
+export function isRunnable(language: string) {
+  return supportedLanguages.includes(language);
+}
+
+/**
+ * Run JavaScript code in eval() context, to support returning values from simple expressions `1+1`
+ * and also support `import * as esbuild from 'https://cdn.skypack.dev/esbuild-wasm@0.19.2'` via ES6 modules fallback
+ */
+async function runJavascript(code: string) {
+  try {
+    const fn = new Function(`return eval(${JSON.stringify(code)});`);
+    return fn();
+  } catch (error: any) {
+    const msgLower = error.message.toLowerCase();
+    const maybeES6Module =
+      error instanceof SyntaxError &&
+      // Cannot use import statement outside a module
+      // import declarations may only appear at top level of a module
+      msgLower.includes("module") &&
+      msgLower.includes("import");
+    if (maybeES6Module) {
+      // check if code has export.*default regexp
+      if (!/export\s+default\s+/.test(code)) {
+        console.warn(
+          "Chatcraft: Evaling code in a module context, must `export default <your val>` at end to return a value"
+        );
+      }
+      const module = await import("data:text/javascript;base64," + btoa(code) /* @vite-ignore */);
+      return module.default;
+    } else {
+      throw error;
+    }
+  }
+}
+
+async function compileWithEsbuild(tsCode: string) {
+  try {
+    await esbuild.initialize({
+      // no idea how to load the bundled wasm file from esbuild-wasm in vite
+      wasmURL:
+        "../../node_modules/.pnpm/esbuild-wasm@0.19.2/node_modules/esbuild-wasm/esbuild.wasm",
+    });
+  } catch (error: any) {
```

Instead of silently failing on re-init, why not add a boolean variable to the module, and set it to `true` on success the first time?

```ts
let initialized = false;

// ...

if (!initialzied) {
  await esbuild.initialize(...);
  initialized = true;
}
```

---

I started that way, that still occasionally failed

---

Race condition?  Maybe we need a value of `'uninitialized' | 'initializing' | 'initialized' | 'error'` instead of a boolean here.

---

interesting theory, just need to stick the promise there

---

no dice, i even tried to disable esbuild worker. For some reason the loaded wasm persists across page reloads whereas my module-global variable does not